### PR TITLE
WIKIEDITOR-35: Ability to add CodeMirror instances from external JavaScript by using RequireJS

### PR DESCRIPTION
**Pull request by yflory**

### CODE REVIEW

File: editor-tool-highlighting/src/main/resources/SyntaxHighlighting/SyntaxHighlighting.xml [hunk 1]
```diff
@@ -125,7 +127,7 @@
 ## Only do work if the extension is enabled.
 #if ($configuration.getValue('enabled') == 1)
 
-(function () {
+define([], function () {
```
Comment (cjdelisle) [inline editor-tool-highlighting/src/main/resources/SyntaxHighlighting/SyntaxHighlighting.xml]: is the [] required ?   I thought `define(function () { ... })` was enough

**Comment by Enygma2002** [comment]
When making a PR, be careful to keep it on topic and corresponding to a particular issue. PRs are not a tool to dump batches of code into the main branch, but a tool to review specific/individual changes that are done to the code in order to fix a specific issue.

You need an individual PR per issue that you are trying to fix.

File: editor-tool-highlighting/pom.xml [hunk 2]
```diff
@@ -45,7 +45,7 @@
     <dependency>
       <groupId>org.webjars</groupId>
       <artifactId>codemirror</artifactId>
-      <version>5.1</version>
+      <version>5.8</version>
```
Comment (vmassol) [inline editor-tool-highlighting/pom.xml]: could you explain how this version upgrade is related to WIKIEDITOR-35?

File: editor-tool-highlighting/src/main/resources/SyntaxHighlighting/SyntaxHighlighting.xml [hunk 3]
```diff
@@ -116,8 +114,12 @@
         <classType>com.xpn.xwiki.objects.classes.StaticListClass</classType>
       </use>
     </class>
+    <name>SyntaxHighlighting.SyntaxHighlighting</name>
+    <number>0</number>
+    <className>XWiki.JavaScriptExtension</className>
+    <guid>ae324f77-9a39-4a22-b789-fb02e302d54f</guid>
     <property>
-      <cache>long</cache>
+      <cache>default</cache>
```
Comment (vmassol) [inline editor-tool-highlighting/src/main/resources/SyntaxHighlighting/SyntaxHighlighting.xml]: no more caching?

File: editor-tool-highlighting/src/main/resources/SyntaxHighlighting/SyntaxHighlighting.xml [hunk 4]
```diff
@@ -299,10 +317,20 @@
               var fullScreenDeactivatorElement = $$('.bottombuttons')[0].down('.button');
               fullScreenDeactivatorElement.click();
             }
-          }
+          },
+          // Enable persistent search
+          'Alt-F': 'findPersistent',
```
Comment (vmassol) [inline editor-tool-highlighting/src/main/resources/SyntaxHighlighting/SyntaxHighlighting.xml]: hmm how is this related to WIKIEDITOR-35?

File: editor-tool-highlighting/src/main/resources/SyntaxHighlighting/SyntaxHighlighting.xml [hunk 5]
```diff
@@ -299,10 +317,20 @@
...
         }
       });
 
+      if(Object.keys(shortcut.all_shortcuts).indexOf('ctrl+g') &gt; 0) {
+        var callback = shortcut.all_shortcuts['ctrl+g'].callback;
+        shortcut.remove('ctrl+g');
+        shortcut.add('ctrl+g', callback, {
+          'disable_in_input' : true
+        });
```
Comment (vmassol) [inline editor-tool-highlighting/src/main/resources/SyntaxHighlighting/SyntaxHighlighting.xml]: same question: hmm how is this related to WIKIEDITOR-35?

**Comment by cjdelisle** [comment]
```You need an individual PR per issue that you are trying to fix.``` there is a valid reason to handle multiple issues in one PR, it is not possible to add commits which build on top of one another in separate PRs otherwise all PRs after the first will duplicate the same commits.

**Comment by vmassol** [comment]
@caleb: the comment for this PR is: "WIKIEDITOR-35: Ability to add CodeMirror instances from external JavaScript by using RequireJS" and a lot of the code is absolutely not related to this. So even if there's only one commit you need a comment that reflects what is in the PR, possibly listing several jira issues in the commit comment.

Also the bigger the PR the harder it is to get accepted. I'm personally unsure about the changes of keyboard shortcuts and I'd need to know more.

**Comment by Enygma2002** [comment]
I had to make the above comments on each commit, instead of making them on the PR's page (`Files changed`) because it would not have made much sense to talk about 3-4 issues in the same view, at the same time. Even Github is built this way (one PR per issue) and yes, @cjdelisle I understand the case you are pointing out, but this was not it. All 3 issues were completely independent issues that can be committed in individual branches, PR-ed and merged automagically.
